tpmutil/run_windows.go

```diff
@@ -0,0 +1,176 @@
+	0x80284014: "The Physical Presence Interface of this firmware does not support the requested method.",
+	0x80284015: "The requested TPM OwnerAuth value was not found.",
+}
+
+func tbsError(err uintptr) error {
+	if err == 0 {
+		return nil
+	}
+	if description, ok := errMap[err]; ok {
+		return fmt.Errorf("TBS Error %v: %s", err, description)
+	}
+	return fmt.Errorf("Unrecognized TBS Error %v", err)
+}
+
+// winTPMBuffer is a ReadWriteCloser to access the TPM in Windows.
+type winTPMBuffer struct {
+	context  uintptr
+	outBuffer []byte
+	priority  CommandPriority
+}
+
+// Executes the TPM command specified by commandBuffer, returning the number of bytes in the command
+// and any error code returned by executing the TPM command. Command response can be read by calling
+// Read().
+func (rwc *winTPMBuffer) Write(commandBuffer []byte) (int, error) {
+	// TPM spec defines longest possible response to be maxTPMResponse.
+	outBufferLen := maxTPMResponse
+	rwc.outBuffer = rwc.outBuffer[:outBufferLen]
+
+	// TBS_RESULT Tbsip_Submit_Command(
+	//   _In_          TBS_HCONTEXT         hContext,
+	//   _In_          TBS_COMMAND_LOCALITY Locality,
+	//   _In_          TBS_COMMAND_PRIORITY Priority,
+	//   _In_    const PCBYTE               *pabCommand,
+	//   _In_          UINT32               cbCommand,
+	//   _Out_         PBYTE                *pabResult,
+	//   _Inout_       UINT32               *pcbOutput
+	// );
+	errResp, _, _ := tbsSubmitCommand.Call(
+		rwc.context,
+		tbsCommandLocalityZero, // Windows currently only supports TBS_COMMAND_LOCALITY_ZERO.
+		uintptr(rwc.priority),
+		uintptr(unsafe.Pointer(&(commandBuffer[0]))),
+		uintptr(len(commandBuffer)),
+		uintptr(unsafe.Pointer(&(rwc.outBuffer[0]))),
+		uintptr(unsafe.Pointer(&outBufferLen)),
+	)
+
+	// Shrink outBuffer so it is length of response.
+	rwc.outBuffer = rwc.outBuffer[:outBufferLen]
+	return len(commandBuffer), tbsError(errResp)
+}
+
+// Provides TPM response from the command called in the last Write call.
+func (rwc *winTPMBuffer) Read(responseBuffer []byte) (int, error) {
+	if len(rwc.outBuffer) == 0 {
+		return 0, io.EOF
+	}
+	lenCopied := copy(responseBuffer, rwc.outBuffer)
+	// Implements same behavior as linux "/dev/tpm0": discard unread components after read.
+	rwc.outBuffer = rwc.outBuffer[:0]
+	return lenCopied, nil
+}
+
+func (rwc *winTPMBuffer) Close() error {
+	// TBS_RESULT Tbsip_Context_Close(
+	//   _In_ TBS_HCONTEXT hContext
+	// );
+	errResp, _, _ := tbsContextClose.Call(rwc.context)
+	return tbsError(errResp)
+}
+
+// OpenTPM creates a new instance of a ReadWriteCloser which can interact with a
+// Windows TPM. OpenTPM takes in the a CommandPriority at which to run commands.
+func OpenTPM(commandPriority CommandPriority) (io.ReadWriteCloser, error) {
+	params := tbsContextParams2{
+		version: tpm2Version,
+		flags:   bothTPMVersionsFlag,
+	}
+
+	rwc := winTPMBuffer{
+		outBuffer: make([]byte, maxTPMResponse),
```

Comment:
`make([]byte, 0, maxTPMResponse)` that way calling Read after OpenTPM returns nothing

Reply:
Done.